Query history: Add search for query history and starred queries

ivanahuckova opened the pull request:
**What this PR does / why we need it**:
Adds search for query history and starred queries. 

The search works on:
- queries
- comments

The search doesn't work on data source name, as we can use data source filter for this. 

Design doc: https://docs.google.com/document/d/1IprNuMA2caFahtEhsDT7Se2aZvIMIF3PldM0kGe-8u8/edit

Images: 
![image](https://user-images.githubusercontent.com/30407135/85304489-1a7d1100-b4ac-11ea-9ff5-b7abc5137b42.png)


![image](https://user-images.githubusercontent.com/30407135/85304566-308ad180-b4ac-11ea-8410-310acbb27043.png)


![image](https://user-images.githubusercontent.com/30407135/85304663-48faec00-b4ac-11ea-9080-0915dcb2a95c.png)


![image](https://user-images.githubusercontent.com/30407135/85304733-616b0680-b4ac-11ea-9388-1154c2806402.png)


![image](https://user-images.githubusercontent.com/30407135/85304957-aa22bf80-b4ac-11ea-8d98-d3c222f4d604.png)

ivanahuckova committed:
Add searching functionality to Queries tab

ivanahuckova committed:
Implement search for starred queries

ivanahuckova committed:
Update tests and add test coverage for searching

s-h-a-d-o-w commented on public/app/core/utils/richHistory.ts:
```diff
@@ -329,6 +329,55 @@ export function notEmptyQuery(query: DataQuery) {
+}
+
+export function filterQueriesByTime(queries: RichHistoryQuery[], timeFilter: [number, number]) {
+  const filteredQueries = queries?.filter(
```
No need for `?`, since it's a required argument. (The same goes for everything in `filterQueriesByDataSource` above.)

ivanahuckova replied:
Resolved in https://github.com/grafana/grafana/pull/25747/commits/3313c6b0e430ceeb85beb17670c8b1215518229a

s-h-a-d-o-w commented on public/app/core/utils/richHistory.ts:
```diff
@@ -329,6 +329,55 @@ export function notEmptyQuery(query: DataQuery) {
+  const sortedQueries = sortQueries(filteredQueriesByDsAndSearchFilter, sortOrder);
+  if (timeFilter) {
+    return filterQueriesByTime(sortedQueries, timeFilter);
+  }
```
I think this filter should also happen before sorting because in this order, there is only a 1 in 4 chance that it'll be faster than the other way around.

ivanahuckova replied:
Updated in https://github.com/grafana/grafana/pull/25747/commits/3313c6b0e430ceeb85beb17670c8b1215518229a

s-h-a-d-o-w commented on public/app/core/utils/richHistory.ts:
```diff
@@ -329,6 +329,55 @@ export function notEmptyQuery(query: DataQuery) {
 
+export function filterQueriesBySearchFilter(queries: RichHistoryQuery[], searchFilter: string) {
+  const filteredQueries = queries.filter(query => {
+    const searchInComment = query.comment.includes(searchFilter);
```
I think it's a good practice to have boolean names start with `is`, `has`, `do`, etc. - e.g. `hasMatchingComment`.

From an algorithmic point of view - if it does have a match, you can return from the function early.

s-h-a-d-o-w commented on public/app/core/utils/richHistory.ts:
```diff
@@ -329,6 +329,55 @@ export function notEmptyQuery(query: DataQuery) {
+  const filteredQueries = queries.filter(query => {
+    const searchInComment = query.comment.includes(searchFilter);
+    const searchInQueries = query.queries.filter(query => {
+      const queryHasSearchedValue = Object.values(_.omit(query, 'datasource')).some(value =>
```
I think in addition to `datasource`, it would also be worthwhile to omit other internal props that definitely exist on every `DataQuery`, like `refId`, `key`, etc.

ivanahuckova replied:
Yes, updated and added comment in https://github.com/grafana/grafana/pull/25747/commits/3313c6b0e430ceeb85beb17670c8b1215518229a

s-h-a-d-o-w replied:
Unless you know of some reason why not to, I would argue that all properties of the base `DataQuery` type should be omitted, even e.g. `hide`. I don't know whether it's set on the queries in the query history but I find it conceivable that a user might e.g. mention "false" in their comments and search for it. ;) (And it's not much overhead to add it to the list of omitted props.)

ivanahuckova replied:
Yes. Great catch and point. I have added the rest of the `DataQuery` properties that we don't want to match.

s-h-a-d-o-w commented on public/app/core/utils/richHistory.ts:
```diff
@@ -329,6 +329,55 @@ export function notEmptyQuery(query: DataQuery) {
+      const queryHasSearchedValue = Object.values(_.omit(query, 'datasource')).some(value =>
+        value.toString().includes(searchFilter)
+      );
+      return queryHasSearchedValue;
```
Storing the result instead of returning it straight away seems a bit excessive here. ;) (Could even do an implicit return.)
The same goes for the other filter functions.

ivanahuckova replied:
Resolved in https://github.com/grafana/grafana/pull/25747/commits/3313c6b0e430ceeb85beb17670c8b1215518229a and https://github.com/grafana/grafana/pull/25747/commits/39a6c845d9891251f7dd11827642fcfd4a690b6b

s-h-a-d-o-w commented on public/app/features/explore/RichHistory/RichHistoryQueriesTab.tsx:
```diff
@@ -31,8 +31,10 @@ export interface Props {
+  searchFilter: string;
   onChangeSortOrder: (sortOrder: SortOrder) => void;
   onSelectDatasourceFilters: (value: SelectableValue[] | null) => void;
+  onSearchQuery: (value: string) => void;
```
I'd call this `onSearchFilterChange`, to be consistent with `searchFilter` and it being a change event.

ivanahuckova replied:
Updated in https://github.com/grafana/grafana/pull/25747/commits/3313c6b0e430ceeb85beb17670c8b1215518229a

ivanahuckova commented on public/app/features/explore/RichHistory/RichHistoryStarredTab.tsx:
```diff
@@ -10,21 +10,24 @@ import { stylesFactory, useTheme } from '@grafana/ui';
 
 import { SortOrder } from '../../../core/utils/explore';
-import { sortQueries, createDatasourcesList } from '../../../core/utils/richHistory';
+import { filterQueries, createDatasourcesList } from '../../../core/utils/richHistory';
```
Yes, makes sense. Renaming. 👍

ivanahuckova replied:
Renamed in https://github.com/grafana/grafana/pull/25747/commits/3313c6b0e430ceeb85beb17670c8b1215518229a

ivanahuckova commented on public/app/core/utils/richHistory.test.ts:
```diff
@@ -168,6 +169,25 @@ describe('createDateStringFromTs', () => {
+    const filteredQueries = filterQueries(mock.storedHistory, SortOrder.Ascending, [], 'i do not exist in query');
+    expect(filteredQueries).toHaveLength(0);
+  });
+  it('should filter out all queries based on search filter', () => {
```
Yes 🙈 !

ivanahuckova committed:
Changes based on provided feedback

ivanahuckova committed:
Remove redundant storing of results

ivanahuckova commented:
Thanks @s-h-a-d-o-w for great and valuable feedback. I have incorporated the updates and changes advised by you. :)

s-h-a-d-o-w commented on public/app/core/utils/richHistory.ts:
```diff
@@ -329,6 +329,54 @@ export function notEmptyQuery(query: DataQuery) {
+      )
+    );
+
+    return hasMatchingQueries.length > 0;
```
`hasMatchingQueries` actually isn't a boolean but an array. Besides, you don't even need it. Unless you want to argue that you find it more readable when you split up into filtering and the return expression. In which case I'd say "Alright but then please rename it." ;) (But you definitely don't need `hasMatchingComment`.)

ivanahuckova replied:
I have renamed it to `listOfMatchingQueries`. I have removed hasMatchingComment.

s-h-a-d-o-w commented:
@ivanahuckova I'm glad that you've found it useful! Just one or two minor notes left but otherwise, it all looks good to me. :)

ivanahuckova committed:
Update naming of onSearchFilterChange function in tests

ivanahuckova committed:
Update filterQueriesBySearchFilter function

ivanahuckova committed:
Update filterQueriesBySearchFilter function

s-h-a-d-o-w reviewed (approved):
Alright, I think it's good to go! :slightly_smiling_face: :+1:

aocenas commented on public/app/features/explore/RichHistory/RichHistoryStarredTab.tsx:
```diff
@@ -108,6 +112,15 @@ export function RichHistoryStarredTab(props: Props) {
+              inputClassName="gf-form-input"
+              placeholder="Search queries"
+              value={searchFilter}
+              onChange={onSearchFilterChange}
```
I think the onChange should be debounced so it does not kick in after each keystroke when typing rapidly. Otherwise depending on the number of items and how they are filtered this can be slow and jarring.

ivanahuckova committed:
Debounce search

ivanahuckova committed:
Remove console.log()

s-h-a-d-o-w commented on public/app/features/explore/RichHistory/RichHistoryQueriesTab.tsx:
```diff
@@ -132,50 +137,49 @@ export function RichHistoryQueriesTab(props: Props) {
+
+  const filterAndSortQueriesDebounced = debounce(searchValue => {
+    setFilteredQueries(filterAndSortQueries(queries, sortOrder, listOfDatasourceFilters, searchValue, timeFilter));
+  }, 300);
```
This actually doesn't seem to work (you can try by mashing your keys - it obviously shouldn't update) because the debounced function is recreated on every render. I would recommend using `use-debounce`, it makes debouncing handlers with hooks so easy...
Unless of course you find common debouncing patterns elsewhere in the code base that are just as straight-forward.

ivanahuckova replied:
Yes you are right. 💙 Thanks so much for suggesting `use-debounce`. Fixed, ready for review.

s-h-a-d-o-w replied:
Debouncing looks good now! :)

ivanahuckova replied:
⛹️‍♀️⛹️‍♀️⛹️‍♀️

ivanahuckova committed:
Use useDebouncedCallback

ivanahuckova commented:
⛹️‍♀️⛹️‍♀️⛹️‍♀️

ivanahuckova committed:
Update tests, remove redundant props

zoltanbedi commented:
I think the search is not working properly in the starred tab. https://www.screencast.com/t/3q3saP2Gmj
I'd expect to return only starred queries if I run it under the starred tab and also when removing the search term I'd expect to see only the starred queries.

ivanahuckova committed:
When in starred, search only in starred

ivanahuckova commented:
Great 👀 ! Yes Zoltan, my copy-paste logic error. 🙈 Fixed in https://github.com/grafana/grafana/pull/25747/commits/ad7ba921c03f07a9a1cf2a4161f44bd5cdbd0ec5.

zoltanbedi commented:
One more thing. The slider on the left for queries tab only works if a term is entered in the search input. I'd expect to filter the history with it only moving the slider. Also I'm not a fan for introducing just another dependency for debounce. You can use lodash debounce like this.
```
  const filterAndSortQueriesDebounced = useCallback(
    debounce((searchValue: string) => {
      setFilteredQueries(
        filterAndSortQueries(
          queries,
          sortOrder,
          datasourceFilters?.map(d => d.value),
          searchValue,
          timeFilter
        )
      );
    }, 300),
    [timeFilter, queries, sortOrder, datasourceFilters]
  );
```
Also there is a [useDebounce](https://github.com/streamich/react-use/blob/master/docs/useDebounce.md) hook in the react-use library which is one of our dependency.

ivanahuckova committed:
Add timefilter as dependency

ivanahuckova committed:
Remove redundant dependency, refactor

ivanahuckova commented:
> The slider on the left for queries tab only works if a term is entered in the search input. I'd expect to filter the history with it only moving the slider. 

Yes, added timeFiler to dependency array in https://github.com/grafana/grafana/pull/25747/commits/488df0372858a74bbd6cd0c3162385948e6749f1


>Also I'm not a fan for introducing just another dependency for debounce. 

Very valid point, removed and refactored in https://github.com/grafana/grafana/pull/25747/commits/a662bb9ee98bf75cadb5d3a418ac82dca4a4c78d

s-h-a-d-o-w commented:
@zoltanbedi Good point! I obviously simply missed that we already have `react-use` and haven't wrapped lodash's debounce in `useCallback` before, so that didn't pop into my head. :)

zoltanbedi reviewed (approved):
Good job. lgtm ✨
